be/src/exprs/agg/retention.h

```diff
@@ -0,0 +1,129 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include <cstring>
+#include <limits>
+#include <type_traits>
+
+#include "column/array_column.h"
+#include "column/binary_column.h"
+#include "column/datum.h"
+#include "column/fixed_length_column.h"
+#include "column/type_traits.h"
+#include "column/vectorized_fwd.h"
+#include "exprs/agg/aggregate.h"
+#include "gen_cpp/Data_types.h"
+#include "glog/logging.h"
+#include "gutil/casts.h"
+#include "runtime/mem_pool.h"
+#include "thrift/protocol/TJSONProtocol.h"
+#include "udf/udf_internal.h"
+#include "util/phmap/phmap_dump.h"
+#include "util/slice.h"
+
+namespace starrocks::vectorized {
+struct RetentionState {
+    void merge_array_element(const DatumArray& datum_array) {
+        auto size = datum_array.size();
+        if (!is_initial) {
+            is_initial = true;
+            boolean_vector.resize(size);
+        }
+
+        // All array conditions' size is equal.
+        DCHECK_EQ(boolean_vector.size(), datum_array.size());
+
+        for (int i = 0; i < size; ++i) {
+            // Use union operation between different rows of the same condition.
+            boolean_vector[i] |= datum_array[i].is_null() ? 0 : datum_array[i].get_uint8();
```

Comment:
If the input is of a definite type, using the specific type for the calculation is a better option.

Reply:
> If the input is of a definite type, using the specific type for the calculation is a better option.

It seems that in our ArrayColumn, Datum/DatumArray does not provide an external type interface.

Reply:
```
auto ele_col = array_column->elements_column()
auto offsets = array_column->offsets()->get_data();

if (ele_col->is_nullable()) {

} else {
    size_t offset = offsets->get_data()[row];
    size_t array_size = offsets->get_data()[row + 1] - offset;
    for (size_t i = 0; i < array_size; ++i) {
        uint32_t ele_offset = offset + i;
        // handle down_cast<BooleanColumn*>(ele_col)->get_data()[ele_offset];
    }
}

```